ui/test/nuxeo-document-layout.test.js

```diff
@@ -0,0 +1,232 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import '@webcomponents/html-imports/html-imports.min.js';
+import '../nuxeo-document-layout.js';
+import '../widgets/nuxeo-input.js';
+import '../widgets/nuxeo-textarea';
+import { Polymer } from '@polymer/polymer/polymer-legacy.js';
+import {
+  fixture,
+  flush,
+  isElementVisible,
+  html,
+  waitForEvent,
+  waitForAttrMutation,
+} from '@nuxeo/nuxeo-elements/test/test-helpers';
+import { LayoutBehavior } from '../nuxeo-layout-behavior.js';
+
+// Export Polymer and PolymerElement for 1.x and 2.x compat
+window.Polymer = Polymer;
+
+window.nuxeo.I18n.language = 'en';
+window.nuxeo.I18n.en = window.nuxeo.I18n.en || {};
+
+Nuxeo = Nuxeo || {};
+Nuxeo.LayoutBehavior = LayoutBehavior;
+
+/* eslint-disable no-unused-expressions */
+suite('<nuxeo-document-layout>', () => {
+  let documentLayout;
+
+  const buildDoc = (props) => {
+    const doc = { type: 'Test', uid: '12ae', properties: {} };
+    Object.assign(doc.properties, props);
+    return doc;
+  };
+
+  const buildLayout = async (doc = buildDoc(), layout = 'edit') => {
+    const dl = await fixture(
+      html`
+        <nuxeo-document-layout .document="${doc}" layout="${layout}"></nuxeo-document-layout>
+      `,
+    );
+    if (!dl.element) {
+      await Promise.race([
+        waitForEvent(dl.$.layout, 'element-changed'),
+        waitForAttrMutation(dl.$.layout.$.error, 'hidden', null),
+      ]);
+    }
+    await flush();
+    return dl;
+  };
+
+  const getWidget = (label) => {
+    if (!documentLayout || !documentLayout.element) {
+      return null;
+    }
+    return documentLayout.element.shadowRoot.querySelector(`[role="widget"][label="${label}"]`);
+  };
+
+  test('Should display an error when the layout is not found', async () => {
+    documentLayout = await buildLayout(buildDoc(), 'view'); // we have no such view layout, so it will result in 404
+    expect(documentLayout.$.layout.$.error.hidden).to.be.false;
+    expect(documentLayout.$.layout.$.error.message).to.equal('documentView.layoutNotFound');
+    isElementVisible(documentLayout.$.layout.$.error);
+    expect(documentLayout.element).to.be.undefined;
+  });
+
+  test('Should load a layout', async () => {
+    documentLayout = await buildLayout();
+    // check the layout was loaded
+    expect(documentLayout.element).to.not.be.null;
+    expect(documentLayout.element.tagName).to.equal('NUXEO-TEST-EDIT-LAYOUT');
+    // assert layout integrity
+    expect(documentLayout.element.document).to.equal(documentLayout.document);
```

Comment:
This is one of the tricky usages of `equal`.
In theory, it would be more bullet proof to use `to.deep.equal`. But I believe the equal will be enough since it is the same instance, correct?
More details: https://github.com/chaijs/deep-eql

Reply:
Yes, what we want here is the `equals` and not the `deep.equals`.